cmd/deploy-terraform/terraform.go

```diff
@@ -0,0 +1,183 @@
+	"github.com/opendevstack/ods-pipeline-terraform/internal/kubernetes"
+	"github.com/opendevstack/ods-pipeline-terraform/internal/output"
+)
+
+const (
+	// exit code returned from terraform plan
+	planSuccessNoChangesExitCode   = 0 // (diff empty)
+	planErrorExitCode              = 1
+	planSuccessWithChangesExitCode = 2 // (diff not empty)
+)
+
+func (d *deployTerraform) terraformCmd(args []string, env map[string]string, dir string, outWriter, errWriter io.Writer) error {
+
+	return command.RunInDir(
+		d.terraformBin, args, env, dir, outWriter, errWriter,
+	)
+}
+
+// terraform plan runs the diff and returns whether the plan is in sync.
+// An error is returned when the plan cannot be started or encounters failures
+func (d *deployTerraform) terraformPlanInSync(args []string, env map[string]string, dir string, outWriter, errWriter io.Writer) (bool, error) {
+	return command.RunWithSpecialFailureCode(
+		d.terraformBin, args, env, dir, outWriter, errWriter, planSuccessWithChangesExitCode,
+	)
+}
+
+func (d *deployTerraform) assembleInitWithK8sBackendArgsEnv() (args []string, env map[string]string, sensitive []string, err error) {
+	args = []string{
+		"init",
+	}
+	// this appears to not be working so we need a backend.tf file.
+	backendConfigs := []string{
+		// "type=kubernetes",
+		// fmt.Sprintf("secret_suffix=%s", d.opts.targetEnvironment),
+		// "in_cluster_config=true",
+	}
+	for _, config := range backendConfigs {
+		args = append(args, fmt.Sprintf("-backend-config=%s", strconv.Quote(config)))
+	}
+	commonArgs := d.commonTerraformArgs()
+	env = d.commonTerraformEnv()
+	env["TF_PLUGIN_CACHE_DIR"] = d.pluginCacheDir
+	sensitive = []string{}
+	for k, v := range d.secretEnvVars {
+		env[k] = v
+		sensitive = append(sensitive, v)
+	}
+	return append(args, commonArgs...), env, sensitive, nil
+}
+
+// assemblePlanArgs creates a slice of arguments for "terraform plan".
+func (d *deployTerraform) assemblePlanArgsEnv() (args []string, env map[string]string, sensitive []string, err error) {
+	empty := []string{}
+	emptyEnv := make(map[string]string)
+	args = []string{
+		"plan",
+		"-detailed-exitcode",
+	}
+	planExtraArgs, err := shlex.Split(d.opts.planExtraArgs)
+	if err != nil {
+		return empty, emptyEnv, empty, fmt.Errorf("parse plan-extra-args (%s): %s", d.opts.planExtraArgs, err)
+	}
+	args = append(args, planExtraArgs...)
+	commonArgs := d.commonTerraformPlanApplyArgs()
+	env = d.commonTerraformPlanApplyEnv()
+	sensitive = []string{}
+	for k, v := range d.secretEnvVars {
+		env[k] = v
+		sensitive = append(sensitive, v)
+	}
+	return append(args, commonArgs...), env, sensitive, nil
+}
+
+// assembleApplyArgs creates a slice of arguments for "terraform apply".
+func (d *deployTerraform) assembleApplyArgsEnv() (args []string, env map[string]string, sensitive []string, err error) {
+	empty := []string{}
+	emptyEnv := make(map[string]string)
+	args = []string{
+		"apply",
+		"-auto-approve", // should save plan instead and use it
```

Comment:
Why is it not using a plan file right now?

Reply:
Yeah I was not sure what to do. The plan contains sensitive information. So I would rather not have the plan be uploaded to nexus. Perhaps one could store the plan in a kubernetes secret and then use it for apply. What do you think?

Reply:
Hmm, I was thinking to only save the plan locally but do not do anything with it. But that doesn't make things much better. Maybe best to initially not change the current approach, and instead leave this for later?

Reply:
Let's file an issue for further discussion once it is merged.

Reply:
See issue #5.